amaranth/hdl/_time.py

```diff
@@ -0,0 +1,249 @@
+
+    def __rmul__(self, other):
+        return self * other
+
+    def __truediv__(self, other):
+        if isinstance(other, Period):
+            return self._femtoseconds / other._femtoseconds
+        elif isinstance(other, numbers.Real):
+            return Period(fs = self._femtoseconds / other)
+        else:
+            return NotImplemented
+
+    def __floordiv__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return self._femtoseconds // other._femtoseconds
+
+    def __mod__(self, other):
+        if not isinstance(other, Period):
+            return NotImplemented
+        return Period(fs = self._femtoseconds % other._femtoseconds)
+
+    def __str__(self):
+        return self.__format__("")
+
+    def __format__(self, format_spec):
+        m = re.match(r"^([1-9]\d*)?(.\d+)?( ?)(([munpf]?)s|([kMG]?)Hz)?$", format_spec)
+
+        if m is None:
+            raise ValueError(f"Invalid format specifier '{format_spec}' for object of type 'Period'")
+
+        width, precision, space, unit, s_unit, hz_unit = m.groups()
+
+        if unit is None:
+            if abs(self._femtoseconds) >= 1_000_000_000_000_000:
+                s_unit = ""
+            elif abs(self._femtoseconds) >= 1_000_000_000_000:
+                s_unit = "m"
+            elif abs(self._femtoseconds) >= 1_000_000_000:
+                s_unit = "u"
+            elif abs(self._femtoseconds) >= 1_000_000:
+                s_unit = "n"
+            elif abs(self._femtoseconds) >= 1_000:
+                s_unit = "p"
+            else:
+                s_unit = "f"
+
+            unit = f"{s_unit}s"
+
+        if s_unit is not None:
+            div, digits = {
+                "":  (1_000_000_000_000_000, 15),
+                "m": (1_000_000_000_000, 12),
+                "u": (1_000_000_000, 9),
+                "n": (1_000_000, 6),
+                "p": (1_000, 3),
+                "f": (1, 0),
+            }[s_unit]
+            integer, decimal = divmod(self._femtoseconds, div)
+
+            if precision:
+                precision = int(precision[1:])
+                decimal = round(decimal * 10**(precision - digits))
+                digits = precision
+
+            value = f"{integer}.{decimal:0{digits}}"
+
+            if not precision:
+                value = value.rstrip('0')
+            value = value.rstrip('.')
+
+        else:
+            if hz_unit == "":
+                value = f"{self.hertz:{precision or ''}f}"
+            elif hz_unit == "k":
+                value = f"{self.kilohertz:{precision or ''}f}"
+            elif hz_unit == "M":
+                value = f"{self.megahertz:{precision or ''}f}"
+            elif hz_unit == "G":
+                value = f"{self.gigahertz:{precision or ''}f}"
+
+        str = f"{value}{space}{unit}"
+        if width:
+            str = f"{str:>{width}}"
+
+        return str
+
+    def __repr__(self):
+        return f"Period({self})"
```

Comment:
I'm wondering if this should be made into actually round-trippable syntax, but ... not sure about it

Reply:
`repr(x)` is supposed to be round-trippable, yes; I think that would be nice to have.